cpp/inc/bond/protocol/protobuf_binary_writer.h

```diff
@@ -0,0 +1,579 @@
+
+        explicit ProtobufBinaryWriter(Buffer& output)
+            : _output{ output },
+              _it{ nullptr }
+        {}
+
+        template <typename T>
+        ProtobufBinaryWriter(Counter& output, const ProtobufBinaryWriter<T>& /*pass1*/)
+            : _output{ output },
+              _it{ nullptr }
+        {}
+
+        bool NeedPass0() const
+        {
+            return _it == nullptr;
+        }
+
+        Pass1 WithPass0(Pass0& pass0)
+        {
+            _it = pass0._lengths.begin();
+            return this;
+        }
+
+        void WriteVersion()
+        {}
+
+        //
+        // Write methods
+        //
+        void WriteStructBegin(const Metadata& /*metadata*/, bool base)
+        {
+            if (base)
+            {
+                detail::proto::NotSupportedException("Inheritance");
+            }
+
+            if (!_fields.empty())
+            {
+                const auto& info = _fields.top(std::nothrow);
+
+                if (info.level > 1)
+                {
+                    detail::proto::NotSupportedException("Container nesting");
+                }
+
+                WriteTag(info.has_element
+                    ? info.element.value.tag
+                    : detail::proto::MakeTag(info.field.id, WireType::LengthDelimited));
+            }
+
+            StructLengthBegin();
+        }
+
+        void WriteStructEnd(bool base = false)
+        {
+            BOOST_VERIFY(!base);
+
+            LengthEnd();
+
+            if (!_fields.empty())
+            {
+                auto& info = _fields.top(std::nothrow);
+
+                if (info.has_element && info.element.map_tag != 0)
+                {
+                    KeyValueEnd(info);
+                }
+            }
+        }
+
+        template <typename T>
+        void WriteField(uint16_t id, const Metadata& metadata, const T& value)
+        {
+            FieldInfo::Element elem;
+            elem.encoding = detail::proto::ReadEncoding(get_type_id<T>::value, metadata);
+            elem.tag = detail::proto::MakeTag(
+                id,
+                detail::proto::GetWireType(get_type_id<T>::value, elem.encoding));
+
+            WriteTag(elem.tag);
+            WriteScalar(value, elem);
+        }
+
+        void WriteFieldBegin(BondDataType type, uint16_t id, const Metadata& metadata)
+        {
+            FieldInfo info;
+            info.has_element = false;
+            info.level = 0;
+            info.field.id = id;
+            info.field.metadata = &metadata;
+
+            _fields.push(info);
+        }
+
+        void WriteFieldBegin(BondDataType type, uint16_t id)
+        {
+            WriteFieldBegin(type, id, schema<Unknown>::type::metadata);
+        }
+
+        void WriteFieldEnd()
+        {
+            _fields.pop(std::nothrow);
+        }
+
+        void WriteContainerBegin(uint32_t size, BondDataType type)
+        {
+            BOOST_VERIFY(size != 0);
```

Comment:
We can't write empty containers?

Reply:
We always skip them (yet to verify a case with `vector<blob>` with empty `blob`(s)).